Make rtm connection more reliable

CharlieHess opened the pull request:
* [x] I've read and understood the [Contributing guidelines](./CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](./CODE_OF_CONDUCT.md).
* [x] I've been mindful about doing atomic commits, adding documentation to my changes, not refactoring too much.
* [x] I've a descriptive title and added any useful information for the reviewer. Where appropriate, I've attached a screenshot and/or screencast (gif preferrably).
* [x] I've written tests to cover the new code and functionality included in this PR.
* [x] I've read, agree to, and signed the [Contributor License Agreement (CLA)](https://docs.google.com/a/slack-corp.com/forms/d/1q_w8rlJG_x_xJOoSUMNl7R35rkpA7N6pUkKhfHHMD9c/viewform).

## PR Summary
This PR tries to address unreliable RTM connections, in a few different ways:

* Double the max pong interval.
  * For comparison, the Slack webapp uses a 60 second timeout, so we've still got a lot of room here.
* Calculate the pong latency differently.
  * Noted in #348, we're imposing a 5 second handicap on ourselves.
* Allow other socket messages to stand in for a pong in some cases.
* Move the ping-pong startup to align with `RTM_CONNECTION_OPENED`.
  * Previously it was started in `WS_OPEN` but that seemed inconsistent.
* Change the default retry policy to reconnect forever.
  * The previous implementation would retry five times in five minutes, then throw `UNABLE_TO_RTM_START`, then repeat the process `MAX_CONNECTION_ATTEMPTS` (10 by default). This was a result of us keeping around two different approaches for retries and led to confusion in #245.
  * @l12s noted that we should just rely on the policy specified in `retryConfig` and remove `MAX_CONNECTION_ATTEMPTS`, which is what we're doing now.

## Related Issues
See #348, #245, and #170.

## Test strategy
There are now tests around the entire retry policy:
```
Retry policy
      ✓ should call the rtm.start error callback if attempts exceed retries
      ✓ should disconnect if the error is unrecoverable
      ✓ should disconnect if autoReconnect is set to false
      ✓ should use max attempts and backoff options in place of a retryConfig
      ✓ should support retrying forever with exponential backoff 
```

aoberoi commented on docs/_reference/RTMClient.md:
```diff
@@ -26,8 +26,9 @@ permalink: /reference/RTMClient
     * [.handleWsMessage(wsMsg)](#RTMClient+handleWsMessage)
     * [._handleMessageAck(replyTo, message)](#RTMClient+_handleMessageAck)
     * [.handleWsError(err)](#RTMClient+handleWsError)
-    * [.handleWsClose()](#RTMClient+handleWsClose)
```
were there arguments that were previously undocumented, or actually no arguments?

CharlieHess replied:
previously undocumented.

aoberoi replied:
got it 👍

aoberoi commented on docs/_reference/RTMClient.md:
```diff
@@ -140,14 +139,16 @@ Disconnects from the RTM API.
 <a name="RTMClient+reconnect"></a>
 
 ### rtmClient.reconnect()
+Attempts to reconnect to the websocket by retrying [rtm.start](rtm.start).
```
maybe not be so specific, since it could be an `rtm.connect()` instead of `rtm.start()`, right?

aoberoi commented on lib/clients/rtm/client.js:
```diff
@@ -47,8 +47,6 @@ var wsSocketFn = require('../transports/ws');
  * @param {boolean} opts.autoReconnect Whether or not to automatically reconnect when the connection
  *     closes.
  * @param {boolean} opts.useRtmConnect True to use rtm.connect rather than rtm.start
- * @param {number} opts.maxReconnectionAttempts
```
one concern here: removing these options from the API is backwards-incompatible. can we give some thought to how we might be able to turn these options into a parameterized retry policy on the fly?

aoberoi replied:
we can also mark them `@deprecated`

CharlieHess replied:
good call, I think we can turn them into a retry policy. Important to note that the current retry behavior is to use `FIVE_RETRIES_IN_FIVE_MINUTES` _first_, and only once that has failed do we count it as a "connection attempt" and apply the backoff defined here. In other words, we're already way off-base in terms of applying these variables.

aoberoi commented on lib/clients/rtm/client.js:
```diff
@@ -359,48 +355,19 @@ RTMClient.prototype.disconnect = function disconnect(optErr, optCode) {
 
 /**
- * Attempts to reconnect to the websocket.
+ * Attempts to reconnect to the websocket by retrying {@link rtm.start}.
```
again, might me too specific

CharlieHess commented:
It was a bit of a PITA but I split the RTM tests into separate categories (`Connection logic`, `Retry policy`, and `Message handling`) and added the following tests for the retry policy:

```
Retry policy
      ✓ should call the rtm.start error callback if attempts exceed retries
      ✓ should disconnect if the error is unrecoverable
      ✓ should disconnect if autoReconnect is set to false
      ✓ should use max attempts and backoff options in place of a retryConfig
      ✓ should support retrying forever with exponential backoff 
```

This fixes #245 and maintains backwards compatibility for folks who were using `maxReconnectionAttempts` or `reconnectionBackoff`.

aoberoi commented on lib/clients/rtm/client.js:
```diff
@@ -35,25 +34,28 @@ var ChatFacet = require('../web/facets').ChatFacet;
+ * @param {Boolean} opts.useRtmConnect  True to use rtm.connect rather than rtm.start
+ * @param {Object} opts.retryConfig     The retry policy to use, defaults to forever with exponential
+ * backoff {@see https://github.com/SEAPUNK/node-retry}
+ * @param {Number} opts.maxReconnectionAttempts DEPRECATED: Use retryConfig instead
```
yay!

aoberoi commented on lib/clients/rtm/client.js:
```diff
@@ -433,21 +414,21 @@ RTMClient.prototype.handleWsMessage = function handleWsMessage(wsMsg) {
   if (includes(RTM_CLIENT_INTERNAL_EVENT_TYPES, message.type)) {
     this._handleWsMessageInternal(message.type, message);
   } else {
+    this._maybeKeepAlive(message);
```
@CharlieHess help me understand this, please. `RTM_CLIENT_INTERNAL_EVENT_TYPES` includes `'pong'`, so pong messages will *not* keep the socket connection alive?

aoberoi replied:
nevermind, i think i see how this works now.

CharlieHess replied:
so the idea is, pong messages will still keep it alive, but if we _don't_ receive a pong message because we are being flooded with so many other legitimate messages, we'll treat those messages like a pong and not close the socket

aoberoi commented on test/clients/rtm/connection-logic.js:
```diff
@@ -0,0 +1,162 @@
+      testReconnectionLogic([firstConnection, secondConnection], done);
+    });
+
+    // This is overly complex for what it's trying to test (that a state var is getting toggled),
```
props for fleshing it out though!

aoberoi commented:
words cannot express how excited i am for this change.

@CharlieHess, this one's for you:

![tenor](https://cloud.githubusercontent.com/assets/494795/26475675/fe45e3d6-416d-11e7-835e-eb8a42a2e3e4.gif)
